zerver/tests/test_messages.py

```diff
@@ -3594,21 +3583,59 @@ def test_claim_attachment(self) -> None:
+        self.assertFalse(Message.content_has_attachment('yo http://foo.com'))
+        self.assertTrue(Message.content_has_attachment('yo\n https://staging.zulip.com/user_uploads/'))
+        self.assertTrue(Message.content_has_attachment('yo\n /user_uploads/%s/wEAnI-PEmVmCjo15xxNaQbnj/photo-10.jpg foo' % (
+            zulip_realm.id,)))
+
+    def test_claim_attachment(self) -> None:
+        user_profile = self.example_user('hamlet')
+        dummy_path_ids = self.setup_dummy_attachments(user_profile)
+        dummy_urls = ["http://localhost:9991/user_uploads/{}".format(x) for x in dummy_path_ids]
+
         # Send message referring the attachment
         self.subscribe(user_profile, "Denmark")
 
-        body = ("Some files here ...[zulip.txt](http://localhost:9991/user_uploads/{realm_id}/31/4CBjtTLYZhk66pZrF8hnYGwc/zulip.txt)" +
-                "http://localhost:9991/user_uploads/{realm_id}/31/4CBjtTLYZhk66pZrF8hnYGwc/temp_file.py.... Some more...." +
-                "http://localhost:9991/user_uploads/{realm_id}/31/4CBjtTLYZhk66pZrF8hnYGwc/abc.py").format(realm_id=realm_id)
+        def assert_attachment_claimed(path_id: str, claimed: bool) -> None:
+            attachment = Attachment.objects.get(path_id=path_id)
+            self.assertEqual(attachment.is_claimed(), claimed)
 
+        # This message should claim attachments 1 only because attachment 2
+        # is not being parsed as a link by Bugdown.
+        body = ("Some files here ...[zulip.txt]({})" +
+                "{}.... Some more...." +
+                "{}").format(dummy_urls[0], dummy_urls[1], dummy_urls[1])
         self.send_stream_message(user_profile.email, "Denmark", body, "test")
+        assert_attachment_claimed(dummy_path_ids[0], True)
+        assert_attachment_claimed(dummy_path_ids[1], False)
 
-        for file_name, path_id, size in dummy_files:
-            attachment = Attachment.objects.get(path_id=path_id)
-            self.assertTrue(attachment.is_claimed())
+        # This message tries to claim the third attachment but fails because
+        # Bugdown would not set has_attachments = True here.
+        body = "Link in code: `{}`".format(dummy_urls[2])
+        self.send_stream_message(user_profile.email, "Denmark", body, "test")
+        assert_attachment_claimed(dummy_path_ids[2], False)
 
-class MessageHasKeywordsTest(ZulipTestCase):
-    '''Test for keywords like has_link, has_image, has_attachment'''
+        # Another scenario where we wouldn't parse the link.
+        body = "Link to not parse: .{}.`".format(dummy_urls[2])
+        self.send_stream_message(user_profile.email, "Denmark", body, "test")
+        assert_attachment_claimed(dummy_path_ids[2], False)
+
+        # Finally, claim attachment 3.
+        body = "Link: {}".format(dummy_urls[2])
+        self.send_stream_message(user_profile.email, "Denmark", body, "test")
+        assert_attachment_claimed(dummy_path_ids[2], True)
+
+class MessageHasKeywordsTest(AttachmentTest):
```

Comment:
Subclassing test classes doens't work how you think it does -- this will run `test_basics` twice.  I don't think we really need a separate class for this bundle of tests.